packages/monitor-v2/src/monitor-og/SnapshotVerification.ts

```diff
@@ -417,6 +420,41 @@ const verifyRules = (parsedRules: RulesParameters, proposal: SnapshotProposalGra
   return { verified: true };
 };
 
+// Check if the proposal has been executed before.
+const hasBeenExecuted = async (
+  currentProposal: TransactionsProposedEvent,
+  params: MonitoringParams
+): Promise<boolean> => {
+  // Get all previous proposals with matching transactions and explanation for the same module that were proposed before
+  // the current proposal.
+  const og = await getOgByAddress(params, currentProposal.address);
+  const previousMatchingProposals = (
+    await runQueryFilter<TransactionsProposedEvent>(og, og.filters.TransactionsProposed(), {
+      start: 0,
+      end: currentProposal.blockNumber,
+    })
+  ).filter(
+    (previousProposal) =>
+      previousProposal.args.proposalHash === currentProposal.args.proposalHash &&
+      previousProposal.args.explanation === currentProposal.args.explanation &&
```

Comment:
Is it possible to have a different explanation string/bytes but resolve to the same ipfs document? Put differently, are ipfs hashes case insensitive or do we allow any "wiggle room" in how this explanation is recorded?

I could imagine the hash is commonly communicated in hex that is case insensitive, so we could see explanations `abcd` and `ABCD`. When passed to snapshot's graphql and to ipfs, these both resolve to the same document, but our replay detection code sees that these are different bytes on-chain (and different UTF-8 strings, so they are not seen as "replays").

Does that make sense? Do we have any protection against this?

Reply:
Snapshot uses CID, not the actual hex hash. CIDv0 is case sensitive while CIDv1 is not. Snapshot's graphql queries are case insensitive for both while CIDv0 URL would not be resolved when passed wrong casing.
We might need to do more in-depth evaluation here, e.g. trying to decode back and forth to check if the CID passed in explanation is good.

Reply:
I tested this on CIDv1 and the bot did not catch duplicate when changing casing in the explanation. We should evaluate this more thoroughly in a separate PR